sirepo/runner_daemon/docker_process.py

```diff
@@ -0,0 +1,161 @@
+            _DOCKER.containers.get, name,
+        )
+    except docker.errors.NotFound:
+        pass
+    else:
+        pkdlog('found stale container {}; removing', name)
+        await trio.run_sync_in_worker_thread(old_container.remove)
+
+
+# Helper w/ shared logic for start_report_job and run_extract_job
+async def _make_container(run_dir, working_dir, quoted_bash_cmd, job_type):
+    # We can't just pass the cmd directly to docker because we have to detour
+    # through bash to set up our environment. But we do assume that it's
+    # sufficiently well-behaved that shlex.quote will work.
+    docker_cmd = [
+        '/bin/bash',
+        '-c',
+        '. ~/.bashrc && ' + quoted_bash_cmd,
+    ]
+
+    name = _container_name(run_dir, job_type)
+
+    await _clear_container(name)
+
+    return await trio.run_sync_in_worker_thread(
+        functools.partial(
+            _DOCKER.containers.run,
+            'radiasoft/sirepo',
+            docker_cmd,
+            # Mount the host's working_dir at run_dir inside the container
+            # {host path: {'bind': container path, 'mode': 'rw'}}
+            volumes={str(working_dir): {'bind': str(run_dir), 'mode': 'rw'}},
+            # This path is interpreted inside the container, so it really uses
+            # working_dir as the working dir:
+            working_dir=str(run_dir),
+            name=name,
+            detach=True,
+            init=True,
+        )
+    )
+
+
+async def start_report_job(run_dir, cmd):
+    run_log_path = run_dir.join(template_common.RUN_LOG)
+    quoted_cmd = ' '.join(shlex.quote(c) for c in cmd)
+    quoted_outpath = shlex.quote(str(run_log_path))
+    quoted_bash_cmd = f'{quoted_cmd} >{quoted_outpath} 2>&1'
+
+    # Use a separate working dir, to make sure that the server isn't accessing
+    # the working dir manually
+    working_dir = run_dir + '-working-dir'
+    pkio.unchecked_remove(working_dir)
+    assert not working_dir.exists()
+    shutil.copytree(str(run_dir), str(working_dir))
+
+    container = await _make_container(run_dir, working_dir, quoted_bash_cmd, 'report')
+
+    return _DockerReportJob(container, {'working_dir': str(working_dir)})
+
+
+class _DockerReportJob:
+    def __init__(self, container, backend_info):
+        self._container = container
+        self.backend_info = backend_info
+
+    async def kill(self, grace_period):
+        with contextlib.suppress(docker.errors.NotFound):
+            await trio.run_sync_in_worker_thread(
+                functools.partial(self._container.stop, timeout=grace_period)
+            )
+
+    async def wait(self):
+        result = await _container_wait(self._container)
+        await trio.run_sync_in_worker_thread(self._container.remove)
+        return result['StatusCode']
+
+
+async def run_extract_job(run_dir, cmd, backend_info):
+    # no output redirection here - we want to let the docker daemon collect it
+    quoted_bash_cmd = ' '.join(shlex.quote(c) for c in cmd)
```

Comment:
The extract jobs are designed to do all their communication without using the disk, to avoid a chicken-and-egg problem: if we save the extract job output into the working dir, then we need to run another extract job to get the results from the first extract job! Also, if there are multiple extract jobs, you have to give them unique names and then keep track of which is which. Given that we already need a mechanism to ship back the return code and stdout, I think it's simplest to ship back stderr as well, and then dump it into the log in the sirepo server.